### ISSUE TITLE
State of Flux-native unit testing?

### ISSUE THREAD

**Issue by mkarlesky** [influxdata/flux#3054]
The [announcement for Flux](https://www.influxdata.com/blog/why-were-building-flux-a-new-data-scripting-and-query-language/) included a brief section on the promise of testability:

> It should be “testable.” Queries are code and they should be tested and checked into source control if they’re going to be a part of a long-lived application. Further, individual parts of queries should be testable in isolation. It should be possible for a user to build up a complicated query, but test each part of that separately.

I’ve interpreted this to mean a plan for some sort of Flux-native unit testing framework. I can see work on this within the repository. I have not seen discussion of it elsewhere. Testing does not seem to be addressed in SPEC.md, nor have I found any reference to testability in release notes, community posts, or dev blog posts. I have seen Go-based testing of Flux community contributions, but I'm asking about testing Flux with Flux.

My [question on the community forums](https://community.influxdata.com/t/unit-testing-flux/15384) has yielded replies but no usable answers as of yet.

In short, my questions are:

1. What's the current state of development of Flux-native testing?
1. Is that functionality minimally usable in recent releases? My team is tackling a transition from InfluxDB to InfluxDB 2, and testing support would be awfully nice (it was sorely missed in the original development of our TICKScripts).
1. Is there any documentation or write-ups on this that I've missed?

Apologies that this issue is probably a bit tangential to core repo issues. I simply wasn't sure where to go next.

**Comment by nathanielc** [comment]
@mkarlesky Thanks for asking. I have read through the question on the forum and will answer your questions here and link back to them.

It looks like you are after a testing entry point that is independent of the Go unit test framework. This is indeed planned and the testing code is structured to enable this kind of behavior. That said there isn't really and documentation on how to do this as we intend to expose this more directly in the future.

In the meantime I can explain how the bits fit together for the Go unit test framework and how you could put together your own Flux-native framework with a little effort.

Here it goes: The Flux testing code is singly focused on testing Flux queries and their results. This means we only have functions for asserting behavior of function that interact with tables. For example you could write  a unit test for this simple function:

```
timeSort = (tables=<-) => tables |> sort(columns:["_time"])
```

but you would not be able to write a test for this simple add function:

```
add = (a,b) => a + b
```


With an understanding of what can and cannot be tested let me peel back the layers and show you how you could do this on your own.

First the `testing` package in the Flux stdlib has a few undocumented functions that are critical to making this work. You can find them here https://github.com/influxdata/flux/blob/master/stdlib/testing/testing.flux#L12

The `inspect` and `run` functions are what you need. These functions expect a test case to be passed to them (in the form of a function that returns an object with some specific fields). The test case defines the input data, the function to call on the data and the expected output data. There is also a `test` keyword in the language. The keyword's purpose is to under the hood call the `testing.inspect` and `testing.run` functions. Currently that under the hood behavior is only implemented as part of the Go unit test hooks, (with plans to eventually expose it in native Flux). What that means is that for now you can directly call the `testing.inspect` and `testing.run` functions instead of relying on the `test` keyword behavior to be implemented.

Here is an example test case written in pure Flux https://github.com/influxdata/flux/blob/master/stdlib/strings/replace_test.flux#L39 that defines an input table and output table and the function to apply to the data, (maps over the data and does string replace). 

Using that test case the Go unit test hooks will see the `test` keyword and call `testing.run` to run the test case. You could instead write the Flux code to explicitly call that function. Note that the `test` keywork acts like an assignment in native Flux so you can simply refer to the test case by its name.

```
// skipping all the data test setup....

test _string_replace = () =>
	({input: testing.loadStorage(csv: inData), want: testing.loadMem(csv: outData), fn: t_string_replace})

testing.run(case: _string_replace)
```

The result of running that Flux will either be an error if the test failed or an empty table if it succeeds. 

In fact this is similar to a workflow we follow on the Flux dev team. Often instead of running the Go unit test repeatedly we use the Flux REPL to load up the test case and call `testing.run/inspect` repeatedly.

```
$ go run ./cmd/flux repl
> @path/to/test.flux // This loads the Flux file into the REPL so we can now reference the testing package and the test case defined by the test keyword
> testing.inspect(case: _string_replace) // inspect the test results
```


So why both `testing.inspect` and `testing.run`? In short the only mechanism we have for reporting test failures is to error the whole query, (we will change this eventually). So `testing.inspect` will execute the Flux functions to be tested but will not make any assertions about them. Conversely `testing.run` simply calls `testing.run` but then asserts that the `diff` of the test is empty which if it is not will fail the query and report the error. That process is a bit awkward currently and we have been exploring ways to have better more precise assertions but that is its current state.



Hope this helps, in short you should be able to write out several Flux test files and then write a script to run those files via the REPL or otherwise that explicitly calls `testing.run` or `testing.inspect` as needed.

**Comment by mkarlesky** [comment]
@nathanielc _Yes._ This is exactly what I was hoping for! Thank you so much.

I need to get my hands dirty with what you've explained to fully grasp it, but I see the outlines of the vision.

A few practical followup questions:

1. Just to confirm I believe you're suggesting we could assemble a test suite out of some appropriate scripting around repl-based shell commands. Yes?
1. Given the limits of what can be tested and how assertions are working, what's the right way in this rig to add in old-fashioned printf-style temporary debugging?
1. Is the package mechanism sufficiently developed to separate test code and production code? That is, can we / how do we import our release Flux queries into the context of the test files. I'm guessing the example you linked was simply that — an all-in-one example.

Grab bag of observations,  questions, and wishlist items:

- I see the lack a native testing framework in the original TICK stack as one of its biggest limitations. Conversely, having testing available in Flux is one of Flux / InfluxDB 2's great strengths.
- If there were a SPEC.md-like development doc for Flux testing I'd be happy to drop in reactions and questions.
- Given the opportunity to create new Flux abilities in Go will the envisioned testing approach meld Flux and Go unit testing? For instance, will it be possible to unit test new Flux written in Go in both Go and also in Flux?
- What is the vision for eventually assembling a proper test suite able to run multiple test cases and collect and present results?
- I absolutely appreciate a purely command line-based approach to running tests. It also occurs to me it might be great to run a Flux test suite from within the InfluxDB web interface, presenting the suite results as a tabular query result.

**Comment by nathanielc** [comment]
> Just to confirm I believe you're suggesting we could assemble a test suite out of some appropriate scripting around repl-based shell commands. Yes?

Yes that should work. Heads up the REPL has a long start up time so you might want to structure it such that you only have to call the REPL once for the entire test suite instead of once per test.

> Given the limits of what can be tested and how assertions are working, what's the right way in this rig to add in old-fashioned printf-style temporary debugging?

Use a combination of `yield` statements and the `testing.inspect` function. The REPL will print out the result of all `yield` statements. So you can inject them at any stage in the query/function and see what data passed through. Using `testing.inspect` you can inspect the inputs outputs and diff accordingly.

>Is the package mechanism sufficiently developed to separate test code and production code? That is, can we / how do we import our release Flux queries into the context of the test files. I'm guessing the example you linked was simply that — an all-in-one example.

This one is harder to overcome in the current state. The package system is complete internally but it only understands the `stdlib` there is no way for it to import your code off the filesystem dynamically. This means when you run the REPL all the code has to be in a single file. 

You might be able to script around this for now. As you have probably seen in the stdlib directories we have `_test.flux` files that are separate from the actual code. You could emulate the same behavior but then have your script concatenate the actual code into the top of the test files before running. Its a bit of a hack but the best I can offer ATM.

**Comment by mkarlesky** [comment]
Thank you. These are all very usable answers, and I think we can get where we want to be from here. On the last point I don't see any big hurdle with rigging up one of many templating libraries to squirt production code into a test file.

**Comment by mkarlesky** [comment]
@nathanielc It took us a while to get here, but we are finally starting to build out a test suite such as what we discussed back in July (above). Largely, we've got it figured out. However, there's one odd thing we're confused by.

You linked to https://github.com/influxdata/flux/blob/master/stdlib/strings/replace_test.flux as an example test we could run. We added `testing.inspect(…).diff` to the end of the file and are able to run it! But, we have to strip out the opening line `package strings_test`. With that line in, executing the test case produces one of two bad behaviors:

1. When we execute the test file line by line with `flux repl` we get `Error: package name mismatch "strings_test" != "main"`.
1. When the file is run with `flux execute @<file name>` we get no results of any sort at STDOUT. Presumably, execution crashes related to the above.

Otherwise, apart from some unforeseen bugs or other glitches, we seem to be on our way preparing tested Flux queries we'll be putting into production for the first time later this month.
